File: vcl/java/com/sun/star/vcl/VCLPath.java
/*************************************************************************
 *
 *  $RCSfile$
 *
 *  $Revision$
 *
 *  last change: $Author$ $Date$
 *
 *  The Contents of this file are made available subject to the terms of
 *  either of the following licenses
 *
 *         - GNU General Public License Version 2.1
 *
 *  Patrick Luby, March 2010
 *
 *  GNU General Public License Version 2.1
 *  =============================================
 *  Copyright 2010 Planamesa Inc.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License version 2.1, as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *  MA  02111-1307  USA
 *
 ************************************************************************/

package com.sun.star.vcl;

import java.awt.Shape;
import java.awt.geom.GeneralPath;

/** 
 * The Java class that implements the convenience methods for accessing Java
 * path information.
 * <p>
 * @version 	$Revision$ $Date$
 * @author 	    $Author$
 */
public final class VCLPath {

	/**
	 * Cached path.
	 */
	private GeneralPath path = new GeneralPath();

	/**
	 * Constructs a new <code>VCLPath</code> instance.
	 */
	public VCLPath() {}

	/**
	 * Closes the path.
	 */
	public void closePath() {

		path.closePath();

	}

	/**
	 * Extends a bezier curve to the path.
	 *
	 * @param firstControlX the first control X coordinate
	 * @param firstControlY the first control Y coordinate
	 * @param secondControlX the second control X coordinate
	 * @param secondControlY the second control Y coordinate
	 * @param x the X coordinate to extend a curve to
	 * @param y the Y coordinate to extend a curve to
	 */
	public void curveTo(float firstControlX, float firstControlY, float secondControlX, float secondControlY, float x, float y) {

		path.curveTo(firstControlX, firstControlY, secondControlX, secondControlY, x, y);

	}

	/**
	 * Extends a line in the path.
	 *
	 * @param x the X coordinate to extend a line to
	 * @param y the Y coordinate to extend a line to
	 */
	public void lineTo(float x, float y) {

		path.lineTo(x, y);

	}

	/**
	 * Moves the current point in the path.
	 *
	 * @param x the X coordinate to move to
	 * @param y the Y coordinate to move to
	 */
	public void moveTo(float x, float y) {

		path.moveTo(x, y);

	}

	/**
	 * Returns the path as a shape.
	 *
	 * @return the path as a shape
	 */
	Shape getShape() {

		return path;

	}

}

File: vcl/java/source/java/VCLPath.cxx
/*************************************************************************
 *
 *  $RCSfile$
 *
 *  $Revision$
 *
 *  last change: $Author$ $Date$
 *
 *  The Contents of this file are made available subject to the terms of
 *  either of the following licenses
 *
 *         - GNU General Public License Version 2.1
 *
 *  Patrick Luby, March 2010
 *
 *  GNU General Public License Version 2.1
 *  =============================================
 *  Copyright 2010 Planamesa Inc.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License version 2.1, as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *  MA  02111-1307  USA
 *
 ************************************************************************/

#define _SV_COM_SUN_STAR_VCL_VCLPATH_CXX

#include <com/sun/star/vcl/VCLPath.hxx>

#if !defined USE_NATIVE_WINDOW || !defined USE_NATIVE_VIRTUAL_DEVICE || !defined USE_NATIVE_PRINTING

using namespace vcl;

// ============================================================================

jclass com_sun_star_vcl_VCLPath::theClass = NULL;

// ----------------------------------------------------------------------------

jclass com_sun_star_vcl_VCLPath::getMyClass()
{
	if ( !theClass )
	{
		VCLThreadAttach t;
		if ( !t.pEnv ) return (jclass)NULL;
		jclass tempClass = t.pEnv->FindClass( "com/sun/star/vcl/VCLPath" );
		OSL_ENSURE( tempClass, "Java : FindClass not found!" );
		theClass = (jclass)t.pEnv->NewGlobalRef( tempClass );
	}
	return theClass;
}

// ----------------------------------------------------------------------------

com_sun_star_vcl_VCLPath::com_sun_star_vcl_VCLPath() : java_lang_Object( (jobject)NULL )
{
	static jmethodID mID = NULL;
	VCLThreadAttach t;
	if ( !t.pEnv )
		return;
	if ( !mID )
	{
		char *cSignature = "()V";
		mID = t.pEnv->GetMethodID( getMyClass(), "<init>", cSignature );
	}
	OSL_ENSURE( mID, "Unknown method id!" );
	jobject tempObj;
	tempObj = t.pEnv->NewObject( getMyClass(), mID );
	saveRef( tempObj );
}

// ----------------------------------------------------------------------------

void com_sun_star_vcl_VCLPath::closePath()
{
	static jmethodID mID = NULL;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()V";
			mID = t.pEnv->GetMethodID( getMyClass(), "closePath", cSignature );	
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			t.pEnv->CallNonvirtualVoidMethod( object, getMyClass(), mID );
	}
}

// ----------------------------------------------------------------------------

void com_sun_star_vcl_VCLPath::curveTo( double _par0, double _par1, double _par2, double _par3, double _par4, double _par5 )
{
	static jmethodID mID = NULL;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "(FFFFFF)V";
			mID = t.pEnv->GetMethodID( getMyClass(), "curveTo", cSignature );	
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
		{
			jvalue args[6];
			args[0].f = jfloat( _par0 );
			args[1].f = jfloat( _par1 );
			args[2].f = jfloat( _par2 );
			args[3].f = jfloat( _par3 );
			args[4].f = jfloat( _par4 );
			args[5].f = jfloat( _par5 );
			t.pEnv->CallNonvirtualVoidMethodA( object, getMyClass(), mID, args );
		}
	}
}

// ----------------------------------------------------------------------------

void com_sun_star_vcl_VCLPath::lineTo( double _par0, double _par1 )
{
	static jmethodID mID = NULL;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "(FF)V";
			mID = t.pEnv->GetMethodID( getMyClass(), "lineTo", cSignature );	
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
		{
			jvalue args[2];
			args[0].f = jfloat( _par0 );
			args[1].f = jfloat( _par1 );
			t.pEnv->CallNonvirtualVoidMethodA( object, getMyClass(), mID, args );
		}
	}
}

// ----------------------------------------------------------------------------

void com_sun_star_vcl_VCLPath::moveTo( double _par0, double _par1 )
{
	static jmethodID mID = NULL;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "(FF)V";
			mID = t.pEnv->GetMethodID( getMyClass(), "moveTo", cSignature );	
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
		{
			jvalue args[2];
			args[0].f = jfloat( _par0 );
			args[1].f = jfloat( _par1 );
			t.pEnv->CallNonvirtualVoidMethodA( object, getMyClass(), mID, args );
		}
	}
}

#endif	// !USE_NATIVE_WINDOW || !USE_NATIVE_VIRTUAL_DEVICE || !USE_NATIVE_PRINTING
